### PR TITLE
Vtctld safety improvements.

### PR DESCRIPTION
**Pull request by alainjobart**
Add flags to protect menus. Also document commands better.

### CODE REVIEW

File: web/vtctld2/src/app/shared/flags/shard.flags.ts [hunk 1]
```diff
@@ -7,6 +7,7 @@ export class DeleteShardFlags {
     this.flags['shard_ref'] = new ShardRefFlag(0, 'shard_ref', keyspaceName + '/' + shardName);
     this.flags['shard_ref']['positional'] = true;
     this.flags['recursive'] = new RecursiveFlag(1, 'recursive');
+    this.flags['even_if_serving'] = new EvenIfServingFlag(1, 'even_if_serving');
```
Comment (pivanof) [inline web/vtctld2/src/app/shared/flags/shard.flags.ts]: Is it okay to have the same position as the previous flag?

File: go/vt/vtctld/api.go [hunk 2]
```diff
@@ -31,7 +31,8 @@ import (
 )
 
 var (
-	localCell = flag.String("cell", "", "cell to use")
+	localCell           = flag.String("cell", "", "cell to use")
+	disableTopologyCRUD = flag.Bool("disable_topology_crud", false, "If set, the CRUD topology actions won't be displayed in the vtctld UI.")
```
Comment (alainjobart) [inline go/vt/vtctld/api.go]: I'll add a vtctld_ prefix to it.

For the rest, the problem is the default value of a boolean flag is usually false, and the default meaning should be to display the items, so it has to be a disable_...

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 3]
```diff
@@ -86,6 +85,9 @@ export class ShardComponent implements OnInit, OnDestroy {
         ]}
       );
     }
+    if (this.featuresService.showTopologyCRUD) {
```
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Moved it up, and added a comment, I think that works.

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 4]
```diff
@@ -97,15 +99,17 @@ export class ShardComponent implements OnInit, OnDestroy {
...
+        {label: 'Ignore Health Error', command: (event) => {this.openIgnoreHealthErrorDialog(); }},
         {label: 'Start Slave', command: (event) => {this.openStartSlaveDialog(); }},
         {label: 'Stop Slave', command: (event) => {this.openStopSlaveDialog(); }},
-        {label: 'Ignore Health Error', command: (event) => {this.openIgnoreHealthErrorDialog(); }},
-        {label: 'Delete', command: (event) => {this.openDeleteTabletDialog(); }},
+        {label: 'Delete Tablet', command: (event) => {this.openDeleteTabletDialog(); }},
       ]},
     ];
     if (this.featuresService.activeReparents) {
+      result[1].items.push(
```
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Comment added too, it works I think.

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 5]
```diff
@@ -244,8 +255,9 @@ export class ShardComponent implements OnInit, OnDestroy {
   }
 
   openDeleteTabletDialog() {
-    this.dialogSettings = new DialogSettings('Delete', `Delete ${this.selectedTablet.label}`,
-                                             `Are you sure you want to delete ${this.selectedTablet.label}?`,
+    this.dialogSettings = new DialogSettings('Delete Tablet', `Delete ${this.selectedTablet.label}`,
+                                             `Are you sure you want to delete ${this.selectedTablet.label}?` +
+                                             ` It will be removed from topology.`,
```
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Done.

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 6]
```diff
@@ -273,7 +288,8 @@ export class ShardComponent implements OnInit, OnDestroy {
   }
 
   openSetReadOnlyDialog() {
-    this.dialogSettings = new DialogSettings('Set', `Set ${this.selectedTablet.label} to Read-Only`, '',
+    this.dialogSettings = new DialogSettings('Set', `Set ${this.selectedTablet.label} to Read-Only`,
```
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Done. And below too.

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 7]
```diff
@@ -318,16 +338,18 @@ export class ShardComponent implements OnInit, OnDestroy {
...
-    let flags = new PingTabletFlags(this.selectedTablet.alias).flags;
+    let flags = new IgnoreHealthCheckFlags(this.selectedTablet.alias).flags;
     this.dialogContent = new DialogContent('tablet_alias', flags, {}, undefined, 'IgnoreHealthError');
     this.dialogSettings.toggleModal();
   }
 
   openDemoteMasterDialog() {
-    this.dialogSettings = new DialogSettings('Demote', `Demote ${this.selectedTablet.label}`, '',
+    this.dialogSettings = new DialogSettings('Demote', `Demote ${this.selectedTablet.label}`,
+                                             `Makes a master tablet read-only, and return its replication position. Use with caution.`,
```
Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Added comment.

File: web/vtctld2/src/app/shared/flags/shard.flags.ts (hunk 1 above)
Comment (alainjobart) [inline web/vtctld2/src/app/shared/flags/shard.flags.ts]: Oops.

**Comment by alainjobart** [comment]
@michael-berlin @pivanof 
PTAL so I can import shortly, thanks.

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 8]
```diff
@@ -206,7 +249,8 @@ export class ShardComponent implements OnInit, OnDestroy {
   }
 
   openPlanRepShardDialog() {
-    this.dialogSettings = new DialogSettings('Reparent', `Planned Reparent in Shard ${this.keyspaceName}/${this.shardName}`, '',
+    this.dialogSettings = new DialogSettings('Reparent', `Planned Reparent in Shard ${this.keyspaceName}/${this.shardName}`,
+                                             'Designate a new master. Assumes the old master is still healthy and reachable.',
```
Comment (michael-berlin) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: How about "Failover to a new master"?

That's a much strong language and better implies the consequences than "designate". "designate" sounds like you're marking a tablet as new master, but you're not executing on it and that would be wrong?

Same comment for emergency reparent.

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 9]
```diff
@@ -244,8 +291,9 @@ export class ShardComponent implements OnInit, OnDestroy {
   }
 
   openDeleteTabletDialog() {
-    this.dialogSettings = new DialogSettings('Delete', `Delete ${this.selectedTablet.label}`,
-                                             `Are you sure you want to delete ${this.selectedTablet.label}?`,
+    this.dialogSettings = new DialogSettings('Delete Tablet', `Delete ${this.selectedTablet.label}`,
+                                             `Are you sure you want to delete ${this.selectedTablet.label}?` +
+                                             ` It will be removed from topology (but vttablet and MySQL won't be touched).`,
```
Comment (michael-berlin) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: nit: *the* topology

We're talking about the user's topology and not in general.

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 10]
```diff
@@ -318,16 +374,19 @@ export class ShardComponent implements OnInit, OnDestroy {
   }
 
   openIgnoreHealthErrorDialog() {
-    this.dialogSettings = new DialogSettings('Ignore', `Ignore Health Check for ${this.selectedTablet.label}`, '',
+    this.dialogSettings = new DialogSettings('Ignore', `Ignore Health Check for ${this.selectedTablet.label}`,
+                                             `Make ${this.selectedTablet.label} ignore certain health errors that match a pattern.`,
```
Comment (michael-berlin) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Maybe describe the consequences in more detail?

e.g. talk about the fact that you can force a tablet with an error to be healthy and vtgate will serve from that tablet then

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 11]
```diff
@@ -318,16 +374,19 @@ export class ShardComponent implements OnInit, OnDestroy {
...
-    let flags = new PingTabletFlags(this.selectedTablet.alias).flags;
+    let flags = new IgnoreHealthCheckFlags(this.selectedTablet.alias).flags;
     this.dialogContent = new DialogContent('tablet_alias', flags, {}, undefined, 'IgnoreHealthError');
     this.dialogSettings.toggleModal();
   }
 
   openDemoteMasterDialog() {
-    this.dialogSettings = new DialogSettings('Demote', `Demote ${this.selectedTablet.label}`, '',
+    this.dialogSettings = new DialogSettings('Demote', `Demote ${this.selectedTablet.label}`,
+                                             `Makes a master tablet read-only, waits for all transactions to finish, ` +
```
Comment (michael-berlin) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: Out of curiosity: This means at the end of the command there won't be a master?

Will the slaves be reconfigured automatically?

Should we mention this here?

File: web/vtctld2/src/app/dashboard/shard.component.ts [hunk 12]
```diff
@@ -97,15 +101,19 @@ export class ShardComponent implements OnInit, OnDestroy {
...
       ]},
     ];
     if (this.featuresService.activeReparents) {
+      // Add a couple new items to the 'Change' section.
+      result[1].items.push(
+        {label: 'Set ReadOnly', command: (event) => {this.openSetReadOnlyDialog(); }},
+        {label: 'Set ReadWrite', command: (event) => {this.openSetReadWriteDialog(); }},
+      );
+      // Add an entire section.
       result.push(
```
Comment (michael-berlin) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: This looks confusing to me.

You're pushing the "Reparent" section for a second time here? (line 86 and line 118)

Am I missing something?

I would expect that some of these menu entries are conditional on the tablet type?

Comment (alainjobart) [inline web/vtctld2/src/app/dashboard/shard.component.ts]: One is the shard menu, the other the tablet menu.

Yes, they should be conditional on the tablet type, but I don't have that logic yet. Another PR.

**Comment by michael-berlin** [comment]
LGTM

**Pull request merged**